usr/lib/tpm_stdll/tpm_specific.c

```diff
@@ -193,17 +194,23 @@ CK_RV token_specific_init(STDLL_TokData_t * tokdata, CK_SLOT_ID SlotNumber,
     tokdata->mech_list_len = tpm_mech_list_len;
 
     // if the user specific directory doesn't exist, create it
-    sprintf(path_buf, "%s", get_pk_dir(tokdata, fname));
-    if (stat(path_buf, &statbuf) < 0) {
-        if (mkdir(path_buf, S_IRUSR | S_IWUSR | S_IXUSR) == -1) {
-            TRACE_ERROR("mkdir(%s): %s\n", path_buf, strerror(errno));
+    if (get_pk_dir(tokdata, fname, PATH_MAX) == NULL) {
+        TRACE_ERROR("pk_dir buffer overflow\n");
+        return CKR_FUNCTION_FAILED;
+    }
+    if (stat(fname, &statbuf) < 0) {
+        if (mkdir(fname, S_IRUSR | S_IWUSR | S_IXUSR) == -1) {
+            TRACE_ERROR("mkdir(%s): %s\n", fname, strerror(errno));
             return CKR_FUNCTION_FAILED;
         }
     }
     // now create userdir/TOK_OBJ if it doesn't exist
-    strncat(path_buf, "/", sizeof(path_buf) - (strlen(path_buf) + 1));
-    strncat(path_buf, PK_LITE_OBJ_DIR,
-            sizeof(path_buf) - (strlen(PK_LITE_OBJ_DIR) + 1));
+    snres = snprintf(path_buf, PATH_MAX, "%s/%s", fname, PK_LITE_OBJ_DIR);
+    /* Support for old glibc: snprintf might return -1 */
```

Comment:
can you please remove this comment ? it does not tell me anything.

Reply:
Done, but newer glibc cannot return -1 from snprintf anymore.  That is what the comment is about (also see NOTES in man 3 snprintf).